src/Bicep.LangServer/Completions/PublicRegistryModuleMetadataProvider.cs

```diff
@@ -16,69 +17,124 @@ namespace Bicep.LanguageServer.Providers
+    /// Provider to get modules metadata that we store at a public endpoint.
+    /// This endpoint caches module names and versions of modules available in this github repository - https://github.com/Azure/bicep-registry-modules
     /// </summary>
     public class PublicRegistryModuleMetadataProvider : IPublicRegistryModuleMetadataProvider
     {
-        private const string LiveDataEndpoint = "https://live-data.bicep.azure.com/modulesMetadata";
-        private List<ModuleMetadata> moduleMetadataCache = new List<ModuleMetadata>();
+        private const string LiveDataEndpoint = "https://aka.ms/BicepModulesMetadata";
+        private readonly TimeSpan cacheValidFor = TimeSpan.FromHours(1);
 
-        public async Task Initialize()
+        private ModuleMetadata[] cachedModules = new ModuleMetadata[] { };
+        private bool isQueryingLiveData = false;
+        private object queryingLiveSyncObject = new();
+        private DateTime? lastSuccessfulQuery;
+
+        public PublicRegistryModuleMetadataProvider(bool initializeCache = false)
         {
-            try
+            if (initializeCache)
             {
-                HttpClient httpClient = new HttpClient();
-                var moduleMetadata = await httpClient.GetStringAsync(LiveDataEndpoint);
-                var metadata = JsonConvert.DeserializeObject<List<ModuleMetadata>>(moduleMetadata);
+                this.UpdateCacheAsync(true);
+            }
+        }
 
-                if (metadata is not null)
+        private void UpdateCacheAsync(bool initialDelay = false)
+        {
+            if (!IsCacheExpired() && cachedModules.Any())
+            {
+                return;
+            }
+
+            lock (this.queryingLiveSyncObject)
+            {
+                if (this.isQueryingLiveData)
                 {
-                    moduleMetadataCache = metadata;
+                    return;
                 }
+
+                this.isQueryingLiveData = true;
             }
-            catch (Exception e)
+
+            var _ = Task.Run(async () =>
             {
-                Trace.WriteLine(string.Format("Encountered following exception while intializing MCR modules metadata: {0}", e.Message));
-            }
+                try
+                {
+                    if (initialDelay)
+                    {
+                        await Task.Delay(5000);
+                    }
 
-            Trace.WriteLine("Unable to initialize MCR modules metadata.");
-        }
+                    var modules = await TryGetModulesLive();
 
-        public async Task<IEnumerable<string>> GetModuleNames()
-        {
-            List<string> moduleNames = new List<string>();
+                    this.cachedModules = modules;
+                    this.lastSuccessfulQuery = DateTime.Now;
+                }
+                finally
+                {
+                    // Throttle unsuccessful requests to avoid spamming the endpoint
```

Comment:
I'm wondering if we need a basic exponential backoff for retries? Querying every 5s feels a bit unnecessary if easy to avoid.

Reply:
I considered that, but if they have no Internet, then connect, I want to pick up on the success earlier than that.  Remember that this only retries when they hit Intellisense, not constantly in the background.

I could do a shorter max, say, 2min?